usr/src/uts/common/fs/zfs/zil.c

```diff
@@ -1107,16 +1117,19 @@ zil_lwb_commit(zilog_t *zilog, itx_t *itx, lwb_t *lwb)
 			char *dbuf;
 			int error;
 
-			if (dlen) {
-				ASSERT(itx->itx_wr_state == WR_NEED_COPY);
+			if (itx->itx_wr_state == WR_NEED_COPY) {
 				dbuf = lr_buf + reclen;
-				lrw->lr_common.lrc_reclen += dlen;
+				lrcb->lrc_reclen += dnow;
+				if (lrwb->lr_length > dnow)
+					lrwb->lr_length = dnow;
+				lrw->lr_offset += dnow;
+				lrw->lr_length -= dnow;
```

Comment:
I understand the reluctance to make changes and retest.  I think this review is dragging on because the code, especially in zil_lwb_commit() is difficult to understand raising concerns about being able to maintain and improve on the code in the future.  The code here to manage splitting the blocks is pretty similar to that in zfs_log_write().  The loop and use of local offset and resid variables make it much more obvious what is going on.  Perhaps it would be worth it to rewrite zil_lwb_commit with the inner part to something like this: 

`static lwb_t *
zil_lwb_commit(zilog_t *zilog, itx_t *itx, lwb_t *lwb)
{
        lr_t *lrc = &itx->itx_lr; /* common log record */
        lr_write_t *lrw = (lr_write_t *)lrc;
        lr_t *lrc_orig = lrc;
        uint64_t offset = lrw->lr_offset;
        uint64_t length = lrw->lr_length;
        char *lr_buf;
        uint64_t txg = lrc->lrc_txg;
        uint64_t reclen = lrc->lrc_reclen;
        uint64_t dlen = 0;
        uint64_t dnow;
        uint64_t lwb_sp;

        if (lwb == NULL)
                return (NULL);

        ASSERT(lwb->lwb_buf != NULL);

        if (lrc->lrc_txtype == TX_WRITE && itx->itx_wr_state == WR_NEED_COPY)
                dlen = P2ROUNDUP_TYPED(
                    lrw->lr_length, sizeof (uint64_t), uint64_t);

        zilog->zl_cur_used += dlen;

        zil_lwb_write_init(zilog, lwb);

        while (dlen > 0 || lrc == lrc_orig)  {
                zilog->zl_cur_used += reclen;
                /*
                 * If this record won't fit in the current log block,
                 * start a new one.  For WR_NEED_COPY optimize layout
                 * for minimal number of chunks.
                 */
                lwb_sp = lwb->lwb_sz - lwb->lwb_nused;
                if (reclen > lwb_sp || (reclen + dlen > lwb_sp &&
                    lwb_sp < ZIL_MAX_WASTE_SPACE && (dlen % ZIL_MAX_LOG_DATA
                    == 0 || lwb_sp < reclen + dlen % ZIL_MAX_LOG_DATA))) {
                        lwb = zil_lwb_write_start(zilog, lwb);
                        if (lwb == NULL)
                                return (NULL);
                        zil_lwb_write_init(zilog, lwb);
                        ASSERT(LWB_EMPTY(lwb));
                        lwb_sp = lwb->lwb_sz - lwb->lwb_nused;
                        ASSERT3U(reclen + MIN(dlen, sizeof(uint64_t)),
                            <=, lwb_sp);
                }

                dnow = MIN(dlen, lwb_sp - reclen);
                lr_buf = lwb->lwb_buf + lwb->lwb_nused;
                bcopy(lrc_orig, lr_buf, reclen);
                lrc = (lr_t *)lr_buf;
                lrw = (lr_write_t *)lrc;
                /*
                 * If it's a write, fetch the data or get its blkptr
                 * as appropriate.
                 */
                if (lrc->lrc_txtype == TX_WRITE) {
                        if (txg > spa_freeze_txg(zilog->zl_spa))
                                txg_wait_synced(zilog->zl_dmu_pool, txg);
                        if (itx->itx_wr_state != WR_COPIED) {
                                char *dbuf;
                                int error;

                                if (itx->itx_wr_state == WR_NEED_COPY) {
                                        dbuf = lr_buf + reclen;
                                        lrc->lrc_reclen += dnow;
                                        lrw->lr_length = length;
                                        lrw->lr_offset = offset;
                                        offset += dnow;
                                        length -= dnow;
                                } else {
                                        ASSERT(itx->itx_wr_state == WR_INDIRECT);
                                        dbuf = NULL;
                                }
                                error = zilog->zl_get_data(
                                    itx->itx_private, lrw, dbuf, lwb->lwb_zio);
                                if (error == EIO) {
                                        txg_wait_synced(zilog->zl_dmu_pool, txg);
                                        return (lwb);
                                }
                                if (error != 0) {
                                        ASSERT(error == ENOENT ||
                                            error == EEXIST ||
                                            error == EALREADY);
                                        return (lwb);
                                }
                        }
                }

                /*
                 * We're actually making an entry, so update lrc_seq to be the
                 * log record sequence number.  Note that this is generally not
                 * equal to the itx sequence number because not all transactions
                 * are synchronous, and sometimes spa_sync() gets there first.
                 */
                lrc->lrc_seq = ++zilog->zl_lr_seq; /* we are single threaded */
                lwb->lwb_nused += reclen + dnow;
                lwb->lwb_max_txg = MAX(lwb->lwb_max_txg, txg);
                ASSERT3U(lwb->lwb_nused, <=, lwb->lwb_sz);
                ASSERT0(P2PHASE(lwb->lwb_nused, sizeof (uint64_t)));

                dlen -= dnow;
        }

        return (lwb);
}
`